### PR TITLE
Improve error recovery around lambdas in MemberSemnticModel.GetEnclosingBinderInternal

### LINKED ISSUE #16919: MemberSemnticModel.GetEnclosingBinderInternal has a flaw around error recovery when lambdas are involved
The function binds outer most enclosing lambda and gives up if it is unable to locate a bound node for the inner most enclosing lambda right after that. This results in a binder that is likely to miss a lot of expected context information. I think that instead we should bind all enclosing lambdas, starting with the outer most, in normal mode and then in error recovery mode until we can get bound node for the inner most lambda. If we were doing this we wouldn't get a crash described in https://devdiv.visualstudio.com/DevDiv/_workitems?id=378641, I believe. We could get wrong result from SemanticModel, but not a crash.

We will likely need to figure out what does it mean to bind query lambda for error recovery.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #16919.
@dotnet/roslyn-compiler Please review.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs [hunk 1]
```diff
@@ -32015,6 +32015,368 @@ public static void Main(string[] args)
...
+                                                                                void Local1 (Action<T> onNext = p3 =>
+                                                                                                                {
+                                                                                                                    weakRef.TryGetTarget(out var x);
+                                                                                                                })
+                                                                                {
+                                                                                }
+                                                                            })
+                                            {
+                                            }
+                                        })
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: Is the indenting important in these tests? If not, consider using 4 spaces.

**Comment by cston** [comment]
LGTM

**Comment by AlekseyTs** [comment]
test vsi please

**Comment by AlekseyTs** [comment]
test eta please

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-ide, @dotnet/roslyn-infrastructure Are vsi tests supposed to work in post-dev15 branch? It looks like they are failing with the following error:
```
Compile Error

CreatePkgDef : error : ArgumentException: No Visual Studio registration attribute found in this assembly. [D:\j\workspace\windows_vsi_p---abd55101\Open\src\VisualStudio\Setup.Dependencies\VisualStudioSetup.Dependencies.csproj]
Indication 1 
```

**Comment by jasonmalinowski** [comment]
@AlekseyTs The issue is known and will be fixed once we get the next merge in out of master.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review, need a second sign-off.

**Comment by jasonmalinowski** [comment]
@dotnet-bot retest vsi please. These should be working again.

**Comment by AlekseyTs** [comment]
Test windows_vsi_p0_open_prtest please

**Comment by AlekseyTs** [comment]
Test windows_vsi_p1_open_prtest please.

**Comment by AlekseyTs** [comment]
Test windows_vsi_p3_open_prtest please.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review, need a second sign-off.

**Comment by AlekseyTs** [comment]
Test vsi please

**Comment by AlekseyTs** [comment]
@MeiChin-Tsai For 15.1 sign-off
@jaredpar FYI

**Comment by MeiChin-Tsai** [comment]
Go ahead. thx.

**Pull request merged**